drivers/storage/portworx/component/autopilot.go

```diff
@@ -807,6 +959,51 @@ func (c *autopilot) isOCPUserWorkloadSupported() bool {
 	return *c.isUserWorkloadSupported
 }
 
+func (c *autopilot) refreshTokenSecret(secret *v1.Secret, clusterNamespace string, ownerRef *metav1.OwnerReference) error {
+
+	// update the token when it is half way to expiration
+	secret.Data[AutopilotSaTokenRefreshTimeKey] = []byte(time.Now().UTC().Add(time.Duration(defaultAutoPilotSaTokenExpirationSeconds/2) * time.Second).Format(time.RFC3339))
+	newToken, err := generateAPSaToken(clusterNamespace, defaultAutoPilotSaTokenExpirationSeconds)
+	if err != nil {
+		return err
+	}
+
+	// update the secret with new token
+	secret.Data[core.ServiceAccountTokenKey] = newToken
+	err = k8sutil.CreateOrUpdateSecret(c.k8sClient, secret, ownerRef)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func generateAPSaToken(clusterNamespace string, expirationSeconds int64) ([]byte, error) {
+	tokenRequest := &authv1.TokenRequest{
+		Spec: authv1.TokenRequestSpec{
+			ExpirationSeconds: &expirationSeconds,
```

Comment:
Kubernetes allows limiting this value using a param to kube-apiserver.
https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/
```
--service-account-max-token-expiration duration
 The maximum validity duration of a token created by the service account token issuer. If an otherwise valid TokenRequest with a validity duration larger than this value is requested, a token will be issued with a validity duration of this value.
```
We need to read the actual expiration duration from the tokenResp and set that in the secret instead of relying on the value that we requested.

Reply:
this is taken care with updated code